parcels/particlefile.py

```diff
@@ -158,6 +158,8 @@ def write(self, pset, time, sync=True, deleted_only=False):
 
                 self.idx = np.append(self.idx, np.zeros(len(first_write)))
 
+                if np.isnan(pset[0].dt):
+                    logger.warning("particle dt not defined yet at particle file writing. Writing at time %g will be ignored." % time)
                 for p in pset:
                     if p.dt*p.time <= p.dt*time:  # don't write particles if they haven't started yet
```

Comment:
The alternative is to make this statement (`if p.dt*p.time <= p.dt*time`) more precise.

Remember that the goal is just to prevent writing a particle when they haven't started yet (i.e. when `time < p.time` in time-forward mode, and `time > p.time` in time-backward mode.

However, if p.dt is NaN while pfile.write is called, then it means that this happens _before_ the pset.execute. So in that case we can assume that the user wants the whole particleset to be written?

So what about changing to `if p.dt*p.time <= p.dt*time or np.isnan(p.dt)` and removing the first warning?

Reply:
If we do that it means you could have exports like:
p0: t0, t1, t2, t3, t4
p1: t0,            t3, t4

If p1 is release at t3, then p1(t0) and p1(t3) will show same ideas.

But we don't have to forget that we can as well release particles at t = (t2+t3)/2, so in this case we would have initial position at least. 
But this will be only the case with pfile manual write as described in issue #458.
Not having the initial position of particles released between 2 timesteps is and will stay like this.

So, we go for exporting any p for which `p.dt=nan`? 
At the end, I would say it is simply a choice.

Reply:
I agree that this is simply a choice we need to make. Perhaps the design choice should be
- If pfile is written as part of pset.execute, do as before
- If pfile is written by an explicit call by users, then write out all particles in the pset, independent of what their start_time is

Note also that there may be a complication that if `dt` is not set yet, then `time` may also not be set; in which case we would not know which time to write for the particles. I guess that we should throw an error then?

Reply:
I agree for first point.
For second point, I would like not to save everything, but yes we can include the nan `dt`.
There is no problem with nan times. Indeed, the `p.time` is only used for the check I wrote earlier.

The time at which to write the particle is the function argument `pfile.write(pset, time)`
And actually, this is strange. We should write the particles at their own time. pfile.write should not have a time argument.
In the way we call the function (`pfile.write(pset, pset[0].time)`), it is redundant, but it might be conflicting.
I think we should clean that part. I'll work on that asap.

Reply:
Actually, I think that we do need to have `time` as an argument in `pfile.write()`, particularly for this bit of code
https://github.com/OceanParcels/parcels/blob/39ffc77cfdeb7e664cb2ff044c11c4b815db369a/parcels/particlefile.py#L164
and also for
https://github.com/OceanParcels/parcels/blob/39ffc77cfdeb7e664cb2ff044c11c4b815db369a/parcels/particlefile.py#L150-L151
So I don't think it is redundant, right?

Reply:
Ok, but then shouldn't we have rather `self.time[i, self.idx[i]] = p.time` to avoid
that if all particles are at time 10, and you write pfile.write(pset, time=5) you break everything

https://github.com/OceanParcels/parcels/blob/39ffc77cfdeb7e664cb2ff044c11c4b815db369a/parcels/particlefile.py#L166

Or actually, is it wrong to simply test `p.time == time or np.isnan(p.dt)` instead of ` p.dt*p.time <= p.dt*time` ?